File: frontend/src/services/ptwService.ts
import api from './api'

export interface LaborCategory {
  category: string
  rate: number
  hours: number
}

export interface PTWScenario {
  id: string
  name: string
  opportunity_id?: string
  scenario_type: string
  description?: string
  labor_categories: LaborCategory[]
  overhead_rate: number
  gaa_rate: number
  fee_rate: number
  total_labor_cost: number
  direct_costs: number
  indirect_costs: number
  total_cost: number
  total_price: number
  competitive_position?: string
  created_at: string
}

export const ptwService = {
  create: async (data: Partial<PTWScenario>): Promise<PTWScenario> => {
    const payload: any = {
      name: data.name,
      scenario_type: data.scenario_type || 'base',
      labor_categories: data.labor_categories || [],
      overhead_rate: data.overhead_rate || 0,
      gaa_rate: data.gaa_rate || 0,
      fee_rate: data.fee_rate || 0,
    }
    
    if (data.opportunity_id) payload.opportunity_id = data.opportunity_id
    if (data.description) payload.description = data.description
    if (data.competitive_position) payload.competitive_position = data.competitive_position
    
    const response = await api.post<PTWScenario>('/ptw/scenarios', payload)
    return response.data
  },

  list: async (opportunityId?: string): Promise<{ scenarios: PTWScenario[] }> => {
    const url = opportunityId 
      ? `/ptw/scenarios?opportunity_id=${opportunityId}`
      : '/ptw/scenarios'
    const response = await api.get<{ scenarios: PTWScenario[] }>(url)
    return response.data
  },

  get: async (id: string): Promise<PTWScenario> => {
    const response = await api.get<PTWScenario>(`/ptw/scenarios/${id}`)
    return response.data
  },

  update: async (id: string, data: Partial<PTWScenario>): Promise<PTWScenario> => {
    const payload: any = {}
    
    if (data.name !== undefined) payload.name = data.name
    if (data.opportunity_id !== undefined) payload.opportunity_id = data.opportunity_id || null
    if (data.scenario_type !== undefined) payload.scenario_type = data.scenario_type
    if (data.description !== undefined) payload.description = data.description
    if (data.labor_categories !== undefined) payload.labor_categories = data.labor_categories
    if (data.overhead_rate !== undefined) payload.overhead_rate = data.overhead_rate
    if (data.gaa_rate !== undefined) payload.gaa_rate = data.gaa_rate
    if (data.fee_rate !== undefined) payload.fee_rate = data.fee_rate
    if (data.competitive_position !== undefined) payload.competitive_position = data.competitive_position
    
    const response = await api.put<PTWScenario>(`/ptw/scenarios/${id}`, payload)
    return response.data
  },

  getScenarios: async (opportunityId: string): Promise<{ scenarios: PTWScenario[] }> => {
    const response = await api.get<{ scenarios: PTWScenario[] }>(`/ptw/opportunities/${opportunityId}/scenarios`)
    return response.data
  },
}

File: backend/app/services/ptw_service.py
"""Price-to-Win service"""
from typing import List, Optional, Dict, Any
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, and_
from decimal import Decimal

from app.models.ptw import PTWModel


async def create_ptw_model(
    db: AsyncSession,
    tenant_id: str,
    data: Dict[str, Any],
) -> PTWModel:
    """Create a PTW model"""
    # Calculate totals
    labor_categories = data.get("labor_categories", [])
    total_labor_cost = sum(
        float(cat.get("rate", 0)) * float(cat.get("hours", 0))
        for cat in labor_categories
    )
    
    direct_costs = total_labor_cost + float(data.get("other_direct_costs", 0) or 0)
    
    # Calculate indirects
    overhead_rate = float(data.get("overhead_rate", 0) or 0) / 100
    gaa_rate = float(data.get("gaa_rate", 0) or 0) / 100
    fee_rate = float(data.get("fee_rate", 0) or 0) / 100
    
    overhead_cost = direct_costs * overhead_rate
    gaa_base = direct_costs + overhead_cost
    gaa_cost = gaa_base * gaa_rate
    cost_base = gaa_base + gaa_cost
    fee = cost_base * fee_rate
    
    total_cost = cost_base + fee
    total_price = total_cost  # Could add profit margin
    
    ptw = PTWModel(
        tenant_id=tenant_id,
        opportunity_id=data.get("opportunity_id"),
        name=data["name"],
        scenario_type=data.get("scenario_type", "base"),
        description=data.get("description"),
        labor_categories=labor_categories,
        total_labor_cost=Decimal(str(total_labor_cost)),
        overhead_rate=Decimal(str(data.get("overhead_rate", 0) or 0)),
        gaa_rate=Decimal(str(data.get("gaa_rate", 0) or 0)),
        fee_rate=Decimal(str(data.get("fee_rate", 0) or 0)),
        direct_costs=Decimal(str(direct_costs)),
        indirect_costs=Decimal(str(overhead_cost + gaa_cost)),
        total_cost=Decimal(str(total_cost)),
        total_price=Decimal(str(total_price)),
        competitive_position=data.get("competitive_position"),
        igce_prediction=data.get("igce_prediction"),
        recommendations=data.get("recommendations"),
    )
    
    db.add(ptw)
    await db.commit()
    await db.refresh(ptw)
    return ptw


async def compare_scenarios(
    db: AsyncSession,
    opportunity_id: str,
    tenant_id: str,
) -> List[Dict[str, Any]]:
    """Compare PTW scenarios for an opportunity"""
    result = await db.execute(
        select(PTWModel).where(
            and_(
                PTWModel.opportunity_id == opportunity_id,
                PTWModel.tenant_id == tenant_id,
            )
        )
    )
    scenarios = result.scalars().all()
    
    return [
        {
            "id": s.id,
            "name": s.name,
            "scenario_type": s.scenario_type,
            "total_price": float(s.total_price or 0),
            "total_cost": float(s.total_cost or 0),
            "competitive_position": s.competitive_position,
        }
        for s in scenarios
    ]

